### PR TITLE
2017Q1.feature.newplayer

### PR DESCRIPTION
**Pull request by chrberger**
* This PR contains the rewrite of player that has a global Index, better caching, and that is prepared to have selective fast-forward/fast-backwards.

### CODE REVIEW

**Comment by chrberger** [comment]
@HangYin2015 @olbender @fgiaimo As this change is rather fundamental, I would like to ask all three of you for a review.

File: libopendavinci/include/opendavinci/odtools/player/Player2.h [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+        /**
+         * This class can be used to replay previously recorded data and return
+         * the contained containers to be relayed into a ContainerConference.
+         */
+        class OPENDAVINCI_API Player2 {
+            private:
+                enum {
+                    ONE_MILLISECOND_IN_MICROSECONDS = 1000,
+                    ONE_SECOND_IN_MICROSECONDS = 1000 *ONE_MILLISECOND_IN_MICROSECONDS,
```
Comment (HangYin2015) [inline libopendavinci/include/opendavinci/odtools/player/Player2.h]: A space missing after *

File: libopendavinci/include/opendavinci/odtools/player/Player2.h [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+                // Fields to compute replay throughput for cache management.
+                odcore::data::TimeStamp m_firstTimePointReturningAContainer;
+                uint64_t m_numberOfReturnedContainersInTotal;
+
+                uint32_t m_delay;
+                uint32_t m_correctedDelay;
+
+            private:
+                /**
+                 * This method set the state of the containerCacheFilling thread.
```
Comment (HangYin2015) [inline libopendavinci/include/opendavinci/odtools/player/Player2.h]: set->sets

File: odcockpit/include/plugins/livefeed/LiveFeedWidget.h [hunk 3]
```diff
@@ -99,16 +100,24 @@ namespace cockpit {
...
 
+                    odcore::reflection::Message resolve(odcore::data::Container &c, bool &successfullyMapped);
+
                 private:
                     odcore::base::Mutex m_dataViewMutex;
-                    unique_ptr<QTreeWidget> m_dataView;
+                    QTreeWidget *m_dataView;
+                    map<int32_t, string> m_containerTypeToName;
+                    odcore::base::Mutex m_containerTypeResolvingMutex;
+                    map<string, bool > m_containerTypeResolving;
```
Comment (HangYin2015) [inline odcockpit/include/plugins/livefeed/LiveFeedWidget.h]: map<string, bool> m_containerTypeResolving;

File: odtools/odplayer/testsuites/Player2TestSuite.h [hunk 4]
```diff
@@ -0,0 +1,2463 @@
...
+                }
+
+                Thread::usleepFor(p2.getDelay());
+                counter++;
+            }
+            // Ensure that all containers have been replayed.
+            TS_ASSERT(11 == counter);
+
+            TimeStamp after;
+            cout << "Duration =
```
Comment (HangYin2015) [inline odtools/odplayer/testsuites/Player2TestSuite.h]: This is only used for temporary experiment, right?

File: libopendavinci/include/opendavinci/odtools/player/Player2.h (hunk 1 above)
Comment (chrberger) [reply libopendavinci/include/opendavinci/odtools/player/Player2.h]: Fixed

File: libopendavinci/include/opendavinci/odtools/player/Player2.h (hunk 2 above)
Comment (chrberger) [reply libopendavinci/include/opendavinci/odtools/player/Player2.h]: Fixed

File: odcockpit/include/plugins/livefeed/LiveFeedWidget.h (hunk 3 above)
Comment (chrberger) [reply odcockpit/include/plugins/livefeed/LiveFeedWidget.h]: Fixed

**Review by HangYin2015 — APPROVED** [review]

File: libopendavinci/include/opendavinci/odtools/player/RecMemIndex.h [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                 * and copies the raw memory into the corresponding shared memory.
+                 *
+                 * @return The next container to be replayed.
+                 */
+                odcore::data::Container makeNextRawMemoryEntryAvailable();
+
+            private:
+                /**
+                 * This method initializes the global index where the sample
+                 * time stamps are sorted chronocally and mapped to the
```
Comment (fgiaimo) [inline libopendavinci/include/opendavinci/odtools/player/RecMemIndex.h]: typo on this comment line

**Review by fgiaimo — APPROVED** [review]

**Comment by chrberger** [comment]
@olbender May I get your review & approval as well?

File: libopendavinci/include/opendavinci/odcore/data/TimeStamp.h [hunk 6]
```diff
@@ -37,6 +37,9 @@ namespace odcore {
          */
         class OPENDAVINCI_API TimeStamp : public odcore::data::TimePoint {
             private:
+                friend class Container;
```
Comment (olbender) [inline libopendavinci/include/opendavinci/odcore/data/TimeStamp.h:40]: Should we allow this?

File: libopendavinci/include/opendavinci/odtools/player/Player.h [hunk 7]
```diff
@@ -52,7 +79,7 @@ namespace odtools {
                  *
                  * @param obj Reference to an object of this class.
                  */
-                Player(const Player &/*obj*/);
+                Player(const Player &/*obj*/) = delete;
```
Comment (olbender) [inline libopendavinci/include/opendavinci/odtools/player/Player.h:82]: These can be put as public, I think it's nice to have them directly after the normal constructor, to directly indicate how and how not this object can be created.

File: libopendavinci/src/odcore/data/Container.cpp [hunk 8]
```diff
@@ -269,13 +269,13 @@ namespace odcore {
             m_serializedData.str(rawData);
 
             // Read sent time stamp data.
-            d->read(3, m_sent);
+            d->read(3, m_sent); m_sent.computeHumanReadableRepresentation();
```
Comment (olbender) [inline libopendavinci/src/odcore/data/Container.cpp:272]: I would vote for having one line each.

**Review by olbender — APPROVED** [review]
Looks good, but it's very difficult to really get into detail on the logic due to the large amounts of code.

**Pull request merged**